lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -36,6 +36,54 @@ static bool hasOperandsOutsideOfBlock(Operation *op) {
   });
 }
 
+/// If _all_ of the operation's operands, which are defined outside of the
+/// operation's own block, are ConstantLike, rematerialize them inside of the
+/// operation's block. This relaxes the conservativeness incurred by the
+/// 'hasOperandsOutsideOfBlock' checks.
+static LogicalResult
+materializeConstantOperandsLocally(Operation *op, PatternRewriter &rewriter) {
+  SmallVector<Value> newOperands;
+  newOperands.reserve(op->getNumOperands());
+  bool doOpRewrite = false;
+
+  for (auto operand : op->getOperands()) {
+    auto defOp = operand.getDefiningOp();
+
+    if (!defOp || defOp->getBlock() == op->getBlock()) {
+      // Keep local operands
+      newOperands.push_back(operand);
+      continue;
+    }
+
+    if (!defOp->hasTrait<OpTrait::ConstantLike>())
+      // Optimization will remain blocked, so don't bother changing the op.
+      return failure();
```

Comment:
This looks like it can make changes and then return failure, which should not happen.

Reply:
Ah, for some reason I assumed the rewriter would take care of rolling back the changes if the pattern fails...